src/procedures/proc_common_neighbors.c

```diff
@@ -0,0 +1,316 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "proc_common_neighbors.h"
+
+#include "../util/arr.h"
+#include "../query_ctx.h"
+#include "../graph/graphcontext.h"
+#define BATCH_SIZE 16384
+
+
+//------------------------------------------------------------------------------
+// Node similarity
+// Computes similarity between pairs of nodes
+// Inspecting each node direct neighbors connected by specified relationship type.
+// Similarity score = common_neighbors(a,b) / (out_degree(a) + out_degree(b) - common_neighbors(a,b))
+//------------------------------------------------------------------------------
+
+// CALL algo.commonNeighbors('act') YIELD node1, node2, score
+
+typedef struct {
+	GrB_Index nrows;            // Rows in A.
+	GrB_Index batch_size;       // Number of rows to process.
+	GrB_Matrix A;               // Relationship matrix.
+	GrB_Matrix At;              // Transposed relationship matrix.
+	GrB_Matrix SIM;             // Similarity matrix, SIM[i,j] = #routes between i and j.
+	GrB_Vector DEGREE;          // Out degree, DEGREE[i] = i's out degree.
+	GxB_Scalar thunk;           // Row offset.
+	GrB_Descriptor desc;        // Descriptor.
+	GxB_SelectOp select_op;     // Selects upper triangle.
+	GxB_MatrixTupleIter *iter;  // Matrix iterator.
+	int iteration;              // Iteration.
+	Node node1;                 // Node 1.
+	Node node2;                 // Node 2.
+	Graph *graph;               // Graph object.
+	SIValue *output;            // Array with 6 entries ["node1", node, "node2", node, "score", score].
```

Comment:
Since this is a fixed-size buffer, can we save ourselves an allocation with:
`SIValue output[6];`